pkg/utils/vlantrunk.go

```diff
@@ -0,0 +1,127 @@
+package utils
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"os/exec"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+/*
+vlan trunk support requires OOT i40e driver for Intel 700 series NICs.
+
+FROM - i40e README
+- trunk
+  Supports two operations: add and rem.
+  - add: adds one or more VLAN id into VF VLAN filtering.
+  - rem: removes VLAN ids from the VF VLAN filtering list.
+  Example 1: add multiple VLAN tags, VLANs 2,4,5,10-20, by PF, p1p2, on
+  a selected VF, 1, for filtering, with the sysfs support:
+  #echo add 2,4,5,10-20 > /sys/class/net/p1p2/device/sriov/1/trunk
+  Example 2: remove VLANs 5, 11-13 from PF p1p2 VF 1 with sysfs:
+  #echo rem 5,11-13 > /sys/class/net/p1p2/device/sriov/1/trunk
+  Note: for rem, if VLAN id is not on the VLAN filtering list, the
+  VLAN id will be ignored.
+
+REGEX vaildation string: ^[0-9]+([,\-][0-9]+)*$
+
+tested on following inputs:
+10
+3,5
+3,200
+10-30
+2,4,5
+2,4,5,10-20
+*/
+
+// IsValidTrunkInput checks if trunk input format valid or not
+func IsValidTrunkInput(inStr string) (bool, error) {
+
+	// Sanity check with REGEX validation
+	var validString = regexp.MustCompile(`^[0-9]+([,\-][0-9]+)*$`)
+	if !validString.MatchString(inStr) {
+		return false, fmt.Errorf("invalid input format")
+	}
+
+	// Extract vlan numbers and check range validation
+	entries := strings.Split(inStr, ",")
+	for i := 0; i < len(entries); i++ {
+		if strings.Contains(entries[i], "-") {
+			rng := strings.Split(entries[i], "-")
+			if len(rng) != 2 {
+				return false, fmt.Errorf("invalid range value %s", rng)
+			}
+			rngSt, err := strconv.Atoi(rng[0])
+			if err != nil {
+				return false, fmt.Errorf("failed to parse %s vlan id, start range is incorrect", rng[0])
+			}
+			rngEnd, err := strconv.Atoi(rng[1])
+			if err != nil {
+				return false, fmt.Errorf("failed to parse %s vlan id,, end range is incorrect", rng[1])
+			}
+
+			if !isValidVlan(rngSt) || !isValidVlan(rngEnd) || rngSt >= rngEnd {
+				return false, fmt.Errorf("invalid vlan range %d-%d", rngSt, rngEnd)
+			}
+		} else {
+			vlan, err := strconv.Atoi(entries[i])
+			if err != nil {
+				return false, fmt.Errorf("failed to parse vlan entry %s", entries[i])
+			}
+
+			if !isValidVlan(vlan) {
+				return false, fmt.Errorf("invalid vlan %d", vlan)
+			}
+
+		}
+	}
+
+	return true, nil
+}
+
+func isValidVlan(vlan int) bool {
+	return vlan > 0 && vlan < 4095
+}
+
+var trunkExecCommand = exec.Command
+var trunkFileFormat = "/sys/class/net/%s/device/sriov/%d/trunk"
+
+// AddVlanTrunk add vlan trunk for given VF(vfID) for a PF 'pfName' from value given in 'value' string
+func AddVlanTrunk(pfName string, vfID int, value string) error {
+	var stdout bytes.Buffer
+	trunkFile := fmt.Sprintf(trunkFileFormat, pfName, vfID)
+
+	if _, err := os.Stat(trunkFile); os.IsNotExist(err) {
+		return fmt.Errorf("file %s does not exist", trunkFile)
+	}
+
+	trunkAddCmd := fmt.Sprintf("echo add %s > %s", value, trunkFile)
```

Comment:
This seems vendor-specific. If it is, would it be possible to have an interface that each vendor can easily implement to provide their own way of configuring these kind of things?

Reply:
It seems Mellanox has something similar to vlan trunk feature, it is called VGT+: https://docs.mellanox.com/pages/viewpage.action?pageId=25146819#SingleRootIOVirtualization(SR-IOV)-VirtualGuestTagging(VGT+)

Not sure I understand it correctly, @moshe010 @adrianchiris may be able to comment.

Reply:
yes it vendor spacify and can't be upstream. see decision on mellanox VGT+ [1] which was not accepted. I am  not sure it good to start putting each vendor his proprietary features. How will tested and maintain it? in the kernel malinglist the they suggested to use switchdev mode so maybe we should implementedit here https://github.com/kubevirt/ovs-cni 

[1] -https://www.spinics.net/lists/netdev/msg610091.html

Reply:
its un-clear to me wether or not vendor specific implementations are to be supported in the CNI, but if they do then im +1 on adding an interface for it so each vendor can have its own implementation.

as Moshe said, testing is going to be an issue if we go down this path.

Reply:
I agree, this is NIC specific. I've put the code out here for feedback as we'd discussed in our meeting. From that same meeting our understanding was that mlx didn't have similar feature. If this is supported by others NIC's then yes, an abstraction is needed. That would also require identifying the NIC at runtime. In that case, we are probably talking about abstraction for every methods and invoke the vendor specific implementation, not just for the vlan trunk.

From above link it doesn't sound like VGT+ is similar to vlan trunk. Can someone explain/confirm this?

I think for switchdev mode we need a different CNI plugin altogether, then again not all NICs are switchdev compatible.